symfit/symmip/mip.py

```diff
@@ -0,0 +1,197 @@
+from dataclasses import dataclass, field
+from typing import List, Dict, Protocol
+from functools import cached_property, reduce
+from collections import defaultdict, namedtuple
+import operator
+import itertools
+
+from sympy import Expr, Rel, Eq, Idx, Indexed, Symbol, lambdify
+from sympy.printing.pycode import PythonCodePrinter
+import numpy as np
+
+from symfit.symmip.backend import GurobiBackend
+from symfit.core.fit_results import FitResults
+from symfit.core.support import key2str
+
+
+DummyModel = namedtuple('DummyModel', 'params')
+
+VTYPES = {'binary': 'B', 'integer': 'I', 'real': 'C',} # ToDo: Enum?
+
+
+class MIPBackend(Protocol):
+    @property
+    def model(self):
+        pass
+
+    @property
+    def printer(self) -> PythonCodePrinter:
+        pass
+
+    def add_var(self, *args, **kwargs):
+        pass
+
+    def add_constr(self, *args, **kwargs):
+        pass
+
+    def set_objective(self, *args, **kwargs):
+        pass
+
+    def update(self, *args, **kwargs):
+        pass
+
+    def optimize(self, *args, **kwargs):
+        pass
+
+    @property
+    def objective_value(self):
+        pass
+
+
+@dataclass
+class MIPResult:
+    popt: Dict
+    objective_value: float
+
+    @property
+    def base2indexed(self):
+        res = {v.base.label: v for v in self.popt if isinstance(v, Indexed)}
+        res.update({v.base: v for v in self.popt if isinstance(v, Indexed)})
+        return res
+
+    def __getitem__(self, v):
+        if not isinstance(v, Indexed) and v in self.base2indexed:
+            v = self.base2indexed[v]
+        return self.popt[v]
+
+
+
+@dataclass
+class MIP:
+    objective: Expr = field(default=None)
+    constraints: List[Rel] = field(default_factory=list)
+    backend: MIPBackend = field(default=GurobiBackend)
+    data: Dict = field(default_factory=dict)
+
+    indices: List[Idx] = field(init=False)
+    variables: List[Symbol] = field(init=False)
+    indexed_variables: List[Indexed] = field(init=False)
+
+    def __post_init__(self):
+        # Prepare constraints
+        self.constraints = [c if isinstance(c, Rel) else Eq(c, 0)
+                            for c in self.constraints]
+        free_symbols = reduce(operator.or_,
+                              (c.free_symbols for c in self.constraints),
+                              self.objective.free_symbols if self.objective else set())
```

Comment:
... huh?

Reply:
Just taking the union of some sets of symbols! ;)

Reply:
Oh right. I can see how this works now. At minimal this needs a comment. Also consider using `set.union` instead of `operator.or_`.